File: packages/core/src/rpg_core/storage/ids.py
from __future__ import annotations

import hashlib
import uuid
from pathlib import Path


def new_id(prefix: str) -> str:
    return f"{prefix}_{uuid.uuid4().hex[:12]}"


def document_id_from_hash(content_hash: str) -> str:
    return f"doc_{content_hash[:12]}"


def page_block_id(document_id: str, page_number: int, block_index: int) -> str:
    return f"block_{document_id}_{page_number:03d}_{block_index:03d}"


def chunk_id(document_id: str, page_start: int, index: int) -> str:
    return f"chunk_{document_id}_{page_start:03d}_{index:03d}"


def hash_file(path: Path) -> str:
    digest = hashlib.sha256()
    with path.open("rb") as fh:
        for chunk in iter(lambda: fh.read(65536), b""):
            digest.update(chunk)
    return digest.hexdigest()

File: packages/ingest-clj/src/rpg/ingest/ids.clj
(ns rpg.ingest.ids
  (:import [java.io FileInputStream]
           [java.math BigInteger]
           [java.security MessageDigest]
           [java.util UUID]))

(defn new-id
  "Generate a prefixed id aligned with rpg_core.storage.ids.new_id."
  [prefix]
  (str prefix "_" (subs (.replace (str (UUID/randomUUID)) "-" "") 0 12)))

(defn document-id-from-hash
  [content-hash]
  (str "doc_" (subs content-hash 0 12)))

(defn page-block-id
  [document-id page-number block-index]
  (format "block_%s_%03d_%03d" document-id page-number block-index))

(defn chunk-id
  [document-id page-start index]
  (format "chunk_%s_%03d_%03d" document-id page-start index))

(defn page-id
  [document-id page-number]
  (format "page_%s_%04d" document-id page-number))

(defn hash-file
  [path]
  (let [digest (MessageDigest/getInstance "SHA-256")
        buffer (byte-array 65536)]
    (with-open [in (FileInputStream. (java.io.File. path))]
      (loop []
        (let [n (.read in buffer)]
          (when (pos? n)
            (.update digest buffer 0 n)
            (recur)))))
    (format "%064x" (BigInteger. 1 (.digest digest)))))
